components/cloud/aws/src/kms.rs

```diff
@@ -0,0 +1,279 @@
+            endpoint: config.location.endpoint,
+        })
+    }
+
+    fn new_with_dispatcher<D>(config: Config, dispatcher: D) -> Result<AwsKms>
+    where
+        D: DispatchSignedRequest + Send + Sync + 'static,
+    {
+        let credentials_provider = util::CredentialsProvider::new()?;
+        Self::new_creds_dispatcher(config, dispatcher, credentials_provider)
+    }
+
+    pub fn new(config: Config) -> Result<AwsKms> {
+        let dispatcher = util::new_http_client()?;
+        Self::new_with_dispatcher(config, dispatcher)
+    }
+}
+
+impl KmsProvider for AwsKms {
+    fn name(&self) -> &[u8] {
+        AWS_KMS_VENDOR_NAME
+    }
+
+    // On decrypt failure, the rule is to return WrongMasterKey error in case it is possible that
+    // a wrong master key has been used, or other error otherwise.
+    fn decrypt_data_key(&self, runtime: &mut Runtime, data_key: &EncryptedKey) -> Result<Vec<u8>> {
+        let decrypt_request = DecryptRequest {
+            ciphertext_blob: bytes::Bytes::copy_from_slice(&*data_key),
+            // Use default algorithm SYMMETRIC_DEFAULT.
+            encryption_algorithm: None,
+            // Use key_id encoded in ciphertext.
+            key_id: Some(self.current_key_id.deref().clone()),
+            // Encryption context and grant tokens are not used.
+            encryption_context: None,
+            grant_tokens: None,
+        };
+        let client = &self.client;
+        let decrypt_response = runtime.block_on(retry(|| {
+            client
+                .decrypt(decrypt_request.clone())
+                .map_err(classify_decrypt_error)
+        }))?;
+        Ok(decrypt_response.plaintext.unwrap().as_ref().to_vec())
+    }
+
+    fn generate_data_key(&self, runtime: &mut Runtime) -> Result<DataKeyPair> {
+        let generate_request = GenerateDataKeyRequest {
+            encryption_context: None,
+            grant_tokens: None,
+            key_id: self.current_key_id.deref().clone(),
+            key_spec: Some(AWS_KMS_DATA_KEY_SPEC.to_owned()),
+            number_of_bytes: None,
+        };
+        let client = &self.client;
+        let generate_response = runtime.block_on(retry(|| {
+            client
+                .generate_data_key(generate_request.clone())
+                .map_err(classify_generate_data_key_error)
+        }))?;
+        let ciphertext_key = generate_response.ciphertext_blob.unwrap().as_ref().to_vec();
+        let plaintext_key = generate_response.plaintext.unwrap().as_ref().to_vec();
+        Ok(DataKeyPair {
+            encrypted: EncryptedKey::new(ciphertext_key)?,
+            plaintext: PlainKey::new(plaintext_key)?,
+        })
+    }
+}
+
+fn classify_generate_data_key_error(err: RusotoError<GenerateDataKeyError>) -> Error {
+    if let RusotoError::Service(e) = &err {
+        match &e {
+            GenerateDataKeyError::InvalidKeyUsage(_) => {
+                Error::KmsError(KmsError::WrongMasterKey(err.into()))
```

Comment:
According to rusoto docs, `InvalidKeyUsage` means the `KEY_USAGE` field is incorrect (i.e. not being `ENCRYPT_DECRYPT` in our case), which should not translate to `WrongMasterKey` error.